### DISCUSSION TITLE
Is there a reason why HNSW was not implemented for GPU?

### DISCUSSION THREAD

**Discussion by jeffwillette** [facebookresearch/faiss#3759] [Q&A]
# Summary

I was hoping to run HNSW on GPU, but I found two issues (https://github.com/facebookresearch/faiss/issues/1611 and https://github.com/facebookresearch/faiss/issues/1348) where @mdouze mentions that HNSW is not implemented for GPU's. I am curious if this is because there was no time/motivation to do it, or if there is any reason which makes HNSW hard (not worth it) to implement on GPU?

Thanks.

**Comment by asadoughi** [comment]
The issues referenced are several years old. Since then, training the HNSW graph on GPU is supported via CAGRA. I suggest running a benchmark for your use case if CAGRA search matches the price-performance of CPU-based HNSW search, but my understanding is that for many applications CPU-based HNSW search is preferred.

**Comment by cjnolet** [comment]
cc @divyegala we should definitely document this. While we're at it, I think we need to improve more documentation around the usage of cuVS within FAISS.
